File: frontend/src/types/category.ts
export interface ApiCategory {
  id: number;
  code: string;
  name: string;
  parent_id: number | null;
  description: string;
  sort_order: number;
  is_system: boolean;
  children: ApiCategory[];
  created_at?: string;
  updated_at?: string;
}

export interface Category {
  id: number;
  code: string;
  name: string;
  parentId: number | null;
  description: string;
  sortOrder: number;
  isSystem: boolean;
  children: Category[];
  createdAt?: string;
  updatedAt?: string;
}

export interface CreateCategoryPayload {
  code: string;
  name: string;
  parent_id?: number | null;
  description?: string;
  sort_order?: number;
}

export interface UpdateCategoryPayload {
  code?: string;
  name?: string;
  description?: string;
  sort_order?: number;
}

export interface CategoryBrief {
  id: number;
  code: string;
  name: string;
}

export function normalizeCategory(api: ApiCategory): Category {
  return {
    id: api.id,
    code: api.code,
    name: api.name,
    parentId: api.parent_id,
    description: api.description,
    sortOrder: api.sort_order,
    isSystem: api.is_system,
    children: api.children.map(normalizeCategory),
    createdAt: api.created_at,
    updatedAt: api.updated_at,
  };
}

File: backend/app/models/category.py
from __future__ import annotations
from datetime import datetime
from typing import TYPE_CHECKING, List, Optional
from sqlalchemy import Integer, String, Text, Boolean, ForeignKey, DateTime, Index
from sqlalchemy.orm import Mapped, mapped_column, relationship
from .base import Base

if TYPE_CHECKING:
    from .book import Book


class Category(Base):
    __tablename__ = "categories"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    code: Mapped[str] = mapped_column(String(32), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    parent_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("categories.id"), nullable=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    is_system: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)

    parent: Mapped[Optional["Category"]] = relationship("Category", remote_side="Category.id", back_populates="children")
    children: Mapped[List["Category"]] = relationship("Category", back_populates="parent")
    books: Mapped[List["Book"]] = relationship("Book", back_populates="category")

    __table_args__ = (
        Index("idx_categories_code", "code"),
        Index("idx_categories_parent_id", "parent_id"),
    )
